components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java

```diff
@@ -0,0 +1,188 @@
+     * entered as <k,v> pairs. These parameters will get appended with an "&".
+     *
+     * @param key   Key.
+     * @param value Value.
+     * @return {@link ServiceURLBuilder}.
+     */
+    @Override
+    public ServiceURLBuilder addFragmentAsParams(String key, String value) {
+
+        fragmentParams = new HashMap<>();
+        fragmentParams.put(key, value);
+        return this;
+    }
+
+    @Override
+    public ServiceURL build() throws URLBuilderException {
+
+        String protocol = fetchProtocol();
+        String hostName = fetchHostName();
+        int port = fetchPort();
+        String resolvedUrlContext = resolveUrlPath(urlPaths);
+        String tenantDomain = resolveTenantDomain();
+
+        StringBuilder resolvedUrlStringBuilder = new StringBuilder();
+        if (isTenantQualifiedUrlsEnabled()) {
+            if (StringUtils.isNotBlank(tenantDomain)) {
+                resolvedUrlStringBuilder.append("/t/").append(tenantDomain);
+            }
+        }
+
+        if (StringUtils.isNotBlank(resolvedUrlContext)) {
+            if (resolvedUrlContext.trim().charAt(0) != '/') {
+                resolvedUrlStringBuilder.append("/").append(resolvedUrlContext.trim());
+            } else {
+                resolvedUrlStringBuilder.append(resolvedUrlContext.trim());
+            }
+        }
+        return new ServiceURL(protocol, hostName, port, resolvedUrlStringBuilder.toString(), parameters, fragment,
+                fragmentParams);
+    }
+
+    private String resolveTenantDomain() {
+
+        String tenantDomain = IdentityTenantUtil.getTenantDomainFromContext();
+        if (StringUtils.isBlank(tenantDomain)) {
+            tenantDomain = PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantDomain();
+        }
+        return tenantDomain;
+    }
+
+    private String resolveUrlPath(String[] urlPaths) {
+
+        StringBuilder urlPathBuilder = new StringBuilder();
+        if (ArrayUtils.isNotEmpty(urlPaths)) {
+            for (String path : urlPaths) {
+                urlPathBuilder.append(path).append("/");
+            }
+            if (urlPathBuilder.toString().endsWith("/")) {
+                urlPathBuilder.setLength(urlPathBuilder.length() - 1);
+            }
+        }
+        return urlPathBuilder.toString();
+    }
+
+    private String fetchProtocol() {
+
+        return CarbonUtils.getManagementTransport();
+    }
+
+    private String fetchHostName() throws URLBuilderException {
+
+        String hostName = ServerConfiguration.getInstance().getFirstProperty(IdentityCoreConstants.HOST_NAME);
+        try {
+            if (StringUtils.isBlank(hostName)) {
+                hostName = NetworkUtils.getLocalHostname();
+            }
+        } catch (SocketException e) {
+            throw new URLBuilderException("Error while trying to resolve the hostname from the system", e);
```

Comment:
Better to log unresolvable host in message so that it makes the message more context aware